kpe/src/model/cnn.rs

```diff
@@ -75,82 +57,160 @@ impl CnnModel {
         embeddings: Embeddings,
         valid_mask: ValidMask,
     ) -> Result<Features, ModelError> {
+        if valid_mask.iter().filter(|valid| **valid).count() < Self::KEY_PHRASE_SIZE {
+            return Err(ModelError::NotEnoughWords);
+        }
+
         debug_assert_eq!(embeddings.shape()[1], valid_mask.len());
-        // TODO: check if the zero padding gives the expected results from the cnn
-        let inputs = tvec!(embeddings.collect(valid_mask)?.into());
-        let outputs = self.plan.run(inputs)?;
-        debug_assert!(outputs[0]
-            .to_array_view::<f32>()?
-            .iter()
-            .all(|v| !v.is_infinite() && !v.is_nan()));
-
-        Ok(outputs[0].clone().into())
+        let valid_embeddings = embeddings.collect(valid_mask)?;
+
+        let run_layer =
+            |idx: usize| self.layers[idx].run(valid_embeddings.t().slice(s![NewAxis, .., ..]));
+        let features = concatenate(
+            Axis(1),
+            &[
+                run_layer(0)?.slice(s![0, .., ..]),
```

Comment:
Is this `slice` necessary, given that `run_layer` already slices it?

Reply:
yes, this slices the output of `run_layer` whereas in `run_layer` the input is sliced. we are adding and removing the singleton batch dimension which is needed due to the more general design of the `Conv1D` layer.